File: clients/client-python/gravitino/client/generic_model_version.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
from typing import Dict, List, Optional

from gravitino.api.metadata_object import MetadataObject
from gravitino.api.metadata_objects import MetadataObjects
from gravitino.api.model.model_version import ModelVersion
from gravitino.api.secret.supports_secrets import SupportsSecrets
from gravitino.client.metadata_object_secret_operations import (
    MetadataObjectSecretOperations,
)
from gravitino.dto.audit_dto import AuditDTO
from gravitino.dto.model_version_dto import ModelVersionDTO
from gravitino.name_identifier import NameIdentifier
from gravitino.utils import HTTPClient


class GenericModelVersion(ModelVersion, SupportsSecrets):
    _model_version_dto: ModelVersionDTO
    """The model version DTO object."""

    def __init__(
        self,
        model_version_dto: ModelVersionDTO,
        rest_client: HTTPClient,
        model_full_ident: NameIdentifier,
    ):
        self._model_version_dto = model_version_dto
        model_version_object: MetadataObject = MetadataObjects.of(
            [
                model_full_ident.namespace().level(1),
                model_full_ident.namespace().level(2),
                model_full_ident.name(),
                str(model_version_dto.version()),
            ],
            MetadataObject.Type.MODEL_VERSION,
        )
        self._object_secret_operations = MetadataObjectSecretOperations(
            model_full_ident.namespace().level(0),
            model_version_object,
            rest_client,
        )

    def version(self) -> int:
        return self._model_version_dto.version()

    def comment(self) -> Optional[str]:
        return self._model_version_dto.comment()

    def aliases(self) -> List[str]:
        return self._model_version_dto.aliases()

    def uris(self) -> Dict[str, str]:
        return self._model_version_dto.uris()

    def properties(self) -> Dict[str, str]:
        return self._model_version_dto.properties()

    def audit_info(self) -> AuditDTO:
        return self._model_version_dto.audit_info()

    def support_secrets(self) -> SupportsSecrets:
        return self

    def get_secrets(self) -> Dict[str, str]:
        return self._object_secret_operations.get_secrets()

File: clients/client-java/src/main/java/org/apache/gravitino/client/GenericModelVersion.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.client;

import java.util.Arrays;
import java.util.Map;
import org.apache.gravitino.Audit;
import org.apache.gravitino.MetadataObject;
import org.apache.gravitino.MetadataObjects;
import org.apache.gravitino.NameIdentifier;
import org.apache.gravitino.dto.model.ModelVersionDTO;
import org.apache.gravitino.model.ModelVersion;
import org.apache.gravitino.secret.SupportsSecrets;

class GenericModelVersion implements ModelVersion, SupportsSecrets {

  private final ModelVersionDTO modelVersionDTO;

  private final MetadataObjectSecretOperations objectSecretOperations;

  GenericModelVersion(
      ModelVersionDTO modelVersionDTO, RESTClient restClient, NameIdentifier modelFullIdent) {
    this.modelVersionDTO = modelVersionDTO;
    MetadataObject modelVersionObject =
        MetadataObjects.of(
            Arrays.asList(
                modelFullIdent.namespace().level(1),
                modelFullIdent.namespace().level(2),
                modelFullIdent.name(),
                String.valueOf(modelVersionDTO.version())),
            MetadataObject.Type.MODEL_VERSION);
    this.objectSecretOperations =
        new MetadataObjectSecretOperations(
            modelFullIdent.namespace().level(0), modelVersionObject, restClient);
  }

  @Override
  public int version() {
    return modelVersionDTO.version();
  }

  @Override
  public Map<String, String> uris() {
    return modelVersionDTO.uris();
  }

  @Override
  public String comment() {
    return modelVersionDTO.comment();
  }

  @Override
  public String[] aliases() {
    return modelVersionDTO.aliases();
  }

  @Override
  public Map<String, String> properties() {
    return modelVersionDTO.properties();
  }

  @Override
  public Audit auditInfo() {
    return modelVersionDTO.auditInfo();
  }

  @Override
  public SupportsSecrets supportsSecrets() {
    return this;
  }

  @Override
  public Map<String, String> getSecrets() {
    return objectSecretOperations.getSecrets();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof GenericModelVersion)) {
      return false;
    }
    GenericModelVersion that = (GenericModelVersion) o;
    return modelVersionDTO.equals(that.modelVersionDTO);
  }

  @Override
  public int hashCode() {
    return modelVersionDTO.hashCode();
  }
}
